src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1125,98 +1134,123 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
+                            HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead;
+            _sendRequestTask = async ?
+                _httpClient.SendAsync(_sendRequestMessage, completionOption, _sendRequestCts!.Token) :
+                Task.Run(() => _httpClient.Send(_sendRequestMessage, completionOption, _sendRequestCts!.Token));
 
-                HttpResponseMessage responseMessage = await _sendRequestTask.ConfigureAwait(false);
+            return _sendRequestTask!;
+        }
 
-                HttpWebResponse response = new HttpWebResponse(responseMessage, _requestUri, _cookieContainer);
+        private async Task<WebResponse> HandleResponse(bool async)
+        {
+            _sendRequestTask ??= SendRequest(async);
 
-                int maxSuccessStatusCode = AllowAutoRedirect ? 299 : 399;
-                if ((int)response.StatusCode > maxSuccessStatusCode || (int)response.StatusCode < 200)
+            try
+            {
+                if (_requestStream is not null) // If we didn't flush and end the request stream, do it now, don't wait for dispose.
                 {
-                    throw new WebException(
-                        SR.Format(SR.net_servererror, (int)response.StatusCode, response.StatusDescription),
-                        null,
-                        WebExceptionStatus.ProtocolError,
-                        response);
+                    if (async)
+                    {
+                        await _requestStream!.FlushAsync().ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        _requestStream!.Flush();
+                    }
+                    _requestStream!.EndWriteOnStreamBuffer();
                 }
+            }
+            catch (ObjectDisposedException)
+            {
+                // If the request stream was disposed, don't do anything.
+            }
+
+            HttpResponseMessage responseMessage = await _sendRequestTask!.ConfigureAwait(false);
+
+            HttpWebResponse response = new HttpWebResponse(responseMessage, _requestUri, _cookieContainer);
 
-                return response;
+            int maxSuccessStatusCode = AllowAutoRedirect ? 299 : 399;
+            if ((int)response.StatusCode > maxSuccessStatusCode || (int)response.StatusCode < 200)
+            {
+                throw new WebException(
+                    SR.Format(SR.net_servererror, (int)response.StatusCode, response.StatusDescription),
+                    null,
+                    WebExceptionStatus.ProtocolError,
+                    response);
             }
-            finally
+
+            _sendRequestMessage?.Dispose();
+
+            if (_disposeRequired)
             {
-                if (disposeRequired)
-                {
-                    client?.Dispose();
-                }
+                _httpClient?.Dispose();
+                _httpClient = null;
```

Comment:
This is probably not needed? (why clear this and not `_sendRequestMessage` for example)